### ISSUE TITLE
Demo Project with Executors?

### ISSUE THREAD

**Issue by gabyx** [rnza0u/blaze#8]
Hi, this tool seems very good focused and I like that it is merely a task orchestrator and thats it.

Some question would have interested me:

- Is there a test repo where we can see also some executors beein written in Rust?

- What made you write another build tool ;-) compared to bazel, buck (havent used them, they are complicated). Also there is dagger.io which lets you define the builds similarly and at the same time use it in CI (but jeah its not free).

- In a monorepo sometimes at least in CI, I want to know what components have changed given some Git rev range, to then select which targets are involved. How would you use this with blaze, maybe there needs to be some support for "change-detection", probably the cache invalidation feature could be used for that by feeding it by the output (list of paths in the mono repo) of a user-defined change detection (git, what ever). `blaze what-changed --changed-paths=git-paths.json --yaml` which would output list if targets which need to be built.

- I am interested in probably using it with writing a Nix executor which chooses a corresponding Nix devshell, or so. 

  Currently we went a bit a different way since I did not see this tool (it looks great!). I wrote `quitsh` which is written (sadly in Go lol) which basically only our way of writting only some typed language instead of dumb shell scripts. But jeah as it turns out when you add more and more it will become another stupid build tool, but I dont want to make it into that, so far, here some explanations:

  https://gitlab.com/data-custodian/custodian/-/tree/main/tools/quitsh?ref_type=heads

  So I might use blaze to orchestrate and underneath just use `quitsh` which runs our scripts and also dispatches ober Nix. 

- Is there a way to have custom data in a component configuration? 
`project.json`. I guess it could also be YAML.

- Another question I had is: How do you cache stuff in CI between steps incorporating blaze? 
and how would you generate CI pipelines when adopting blaze, my goal is: CI runs exactly the same as local development (we achieved this so far pretty good wit `quitsh`, since it only runs over Nix. 
Is there a vision for blaze to somehow generate (user code) a Gitlab pipeline or something, bit for that I guess one needs access to the DAG, in Rust or as an output format (Yaml) to tool it by some other tool (because I dont think you want that part in this tool ;))

- Also I am happy to contribute: I could add a Nix `flake.nix` which builds this Rust project and add it to nixpkgs :), as a first starter to get to know a bit the thing. Would be great because to make it more accesible also by other Nix folks in Monorepo builds etc :)

Happy to once chat with you, if you like. 
It would be cool to have a video chat with you about your tool, vision, devops etc. Do you have a matrix account @gabyx:matrix.org.

BR! Nice work!

**Comment by rnza0u** [comment]
Hi

Thanks a lot for your interest in my project. It is really nice and hearth warming !

I will try to answer your questions as good as i can.

> Is there a test repo where we can see also some executors beein written in Rust?

For now, i only wrote Node executors for my personal usage, they are here: https://github.com/rnza0u/blaze-executors

For Rust executors, there is a small example in the documentation: https://github.com/rnza0u/blaze-executors

I didn't really use them for now, but it should be pretty straightforward to write one.

> What made you write another build tool ;-) compared to bazel, buck (havent used them, they are complicated). Also there is dagger.io which lets you define the builds similarly and at the same time use it in CI (but jeah its not free).

I also think that these build tools are very complicated. Here is a little "about" section i wrote that might answer your question: https://blaze-monorepo.dev/about

> In a monorepo sometimes at least in CI, I want to know what components have changed given some Git rev range, to then select which targets are involved. How would you use this with blaze, maybe there needs to be some support for "change-detection", probably the cache invalidation feature could be used for that by feeding it by the output (list of paths in the mono repo) of a user-defined change detection (git, what ever). blaze what-changed --changed-paths=git-paths.json --yaml which would output list if targets which need to be built.

The `cache` feature should be very useful in most purposes. The only thing that must be manually done in CI is to actually store the `.blaze` directory which contains all the cache entries. Also you would need to cache build artifacts and such.

There is also a `commandFails` invalidation strategy that might be used in order to call `git` and check for changes between commits.

> I am interested in probably using it with writing a Nix executor which chooses a corresponding Nix devshell, or so. Currently we went a bit a different way since I did not see this tool (it looks great!). I wrote quitsh which is written (sadly in Go lol) which basically only our way of writting only some typed language instead of dumb shell scripts. But jeah as it turns out when you add more and more it will become another stupid build tool, but I dont want to make it into that, so far, here some explanations:

> https://gitlab.com/data-custodian/custodian/-/tree/main/tools/quitsh?ref_type=heads

> So I might use blaze to orchestrate and underneath just use quitsh which runs our scripts and also dispatches ober Nix.

You are more than welcome to do so ! Blaze is really great for sitting on top of existing tools like packages managers and user scripts.

> Is there a way to have custom data in a component configuration? project.json. I guess it could also be YAML.

You can use JSON, YAML or Jsonnet (which is actually the recommended format). If i understand correctly the `custom data`, you might be talking about variables ? Yes, Blaze has a lot of templating features, including variables, dynamic configuration when using Jsonnet, built-in helpers, and also custom template helpers (for running custom code when rendering the configuration). The documentation is here: https://blaze-monorepo.dev/docs/guides/templating

> Another question I had is: How do you cache stuff in CI between steps incorporating blaze?

I do it manually using a custom Docker image on my CI. Blaze does not dictate how to cache folders between CI runs.

> and how would you generate CI pipelines when adopting blaze, my goal is: CI runs exactly the same as local development (we achieved this so far pretty good wit quitsh, since it only runs over Nix.

Blaze is not really meant to generate CI pipelines. You can make your CI steps easier with Blaze (since it will help orchestrate everything).

Depending on your needed CI flow, you might be able to write a Blaze executor that iterates across all projects within the workspace and generate a CI template from there.

> Is there a vision for blaze to somehow generate (user code) a Gitlab pipeline or something, bit for that I guess one needs access to the DAG, in Rust or as an output format (Yaml) to tool it by some other tool (because I dont think you want that part in this tool ;))

Like i said before, a custom executor could help in doing that, yes. In my vision, the core of Blaze (outside of custom executors) must remain totally agnostic from CI tools.

> Also I am happy to contribute: I could add a Nix flake.nix which builds this Rust project and add it to nixpkgs :), as a first starter to get to know a bit the thing. Would be great because to make it more accesible also by other Nix folks in Monorepo builds etc :)

You would be very welcome to do that ! I don't really know Nix, but that could really boost up the project, indeed.

I am probably the only user right now, so i would be glad to see more people coming in ! There are still a lot of things that could be improved (or re-done), and now i really need more contributors.

> Happy to once chat with you, if you like. It would be cool to have a video chat with you about your tool, vision, devops etc. Do you have a matrix account @gabyx:matrix.org.

I will create an account very soon and reach to you. I would be happy to talk with you about Blaze !

Anyway, thanks a lot, man. And i hope you find this tool useful. cheers

**Comment by gabyx** [comment]
Thanks for the information.

@rnza0u Another question about the caching (because its IMO maybe weirdly named?): 

What `blaze` stores is only the notion about if sepcified inputs (and outputs) have changed right it does not cache the full
output artifacts of targets?

So when `blaze` runs target B which depends on A, 

- and A has already been run which produced a file `Gen.txt`
- and nothing has changed in A's input

then
-  `blaze` does not run target A 
- it does run target B 
  - It does not magically retrieve the `Gen.txt` file from the cache of target A in case `Gen.txt` is not existing, right?

**Comment by rnza0u** [comment]
you are correct. Blaze only stores file metadata. by default it stores file hash and modification time. it will re compute the hash and compare it with the cached hash only if the modification time has changed (`Mixed` strategy).

you can also use the `Timestamps` strategy if you want to only store modification time. or `Hash` strategy if you want to rely on computing hashes only.

please do not hesitate if you want further clarification !
